### DISCUSSION TITLE
Is there a performance penalty for creating new Dexie instances on every use?

### DISCUSSION THREAD

**Discussion by VapidLinus** [dexie/Dexie.js#1656] [Q&A] [answered]
Here to ask a quick question and hoping someone already has this knowledge before I start performing my own tests.

Currently in our prototype, in `useLiveQuery`, my team creates a new `Dexie` instance every time. Actually, several times per query as we have a dynamic list of databases:

![image](https://user-images.githubusercontent.com/4418470/211516020-55c823a3-2b33-4525-83d9-59f6021c3595.png)

Obviously, this is not optimal and we will store/cache the instances in the future but we're short on time - so I'd like to know whether doing it the way we're doing it here is hurting performance so I know whether improving this immediately should be a high priority. Maybe there is some overhead for creating a new instance, or a cache we're missing out on?

Does anyone have knowledge whether creating new instances on every access is having a noticable or major effect on query latency/performance?

Thanks :)

**Comment by dfahlander** [comment, marked as answer]
I advice you to prioritize improving it as soon as possible because each Dexie will spend some time opening the database. Also, this might leak resources and memory as databases aren't properly closed.

Suggesion would be to have a static Map between customerId and Dexie instances, a helper function to get a Dexie instance based on customerId, that get it from that cache instead of creating a new Dexie each time.

**Reply by VapidLinus** [reply]
Thanks for your quick reply and advice that a instance cache was definitely needed asap 🙌

I spent a little time writing a quick global dexie cache using a simple Map.

This was defintely worth doing asap as I already see reduced latency when doing DB queries. It also seems to have fixed a crash that snuck up recently which seemingly randomly caused the current tab to freeze -  I wasn't sure whether this was related but can confirm now that this was also fixed by not creating new dexie instances all the time.
